libflux/Polymorphic_Labels.md

```diff
@@ -0,0 +1,103 @@
+
+```flux
+builtin fill : (<-tables: stream[{ A with 'column: B }], ?column: 'column, ?value: B, ?usePrevious: bool) => stream[{ A with 'column: B }]
+    where A: Record
+```
+
+`'column` in this case indicates a polymorphic label which would allow `fill` to be called as normal, however when passing a "label" (a string) that is known at compile time it lets us enforce the typing of the input and output records.
+
+```flux
+
+// 'column is inferred to `a` which propagates to the input and output records, making them `{ A with a: B }`
+[{ a: 1 }] |> fill(column: "a", value: 0)
+
+// ERROR: the input record lacks the field `a`
+[{}] |> fill(column: "a", value: 0)
+```
+
+
+## Explanation
+
+Polymorphic labels adds the ability to define record types where the field names (labels) can be a variable instead of a literal name in a similar way that type variables can exist in a type.
+
+```
+{ 'col: string } vs { some_column: string }
+(x: A) => A vs (x: string) => string
+```
+
+The types that fit into a record field implement the new `Label` kind to ensure that unexpected types like `Int` are an error when used in a record field.
+
+To make the change as transparent as possible string literals now get the new `Label("literal")` type instead of `String` making `"abc": Literal("abc")`. While this means that two different string literals will have different types we must still treat `Label` types as strings in most cases. Consider
+
+```flux
+if b then "a" else "b"
+```
+
+The branches will have the `Label("a")` and `Label("b")` respectively, which we still need to unify successfully and the resulting type of the expression should still be a `string`. As long as we limit  the uses of label types to direct uses in calls (like in `fill` above). We may keep the same type checking behaviour by treating `Label` as a `String` types in every instance except when type checking function calls, in which case we specialize unification to a subsumption check where we allow some limited sub-typing.
+
+```flux
+// The original fill call is allowed such that `'column` is inferred to be `Label("a")`
+fill(column: "a", value: 0)
+
+// We should (probably) still allow dynamic strings for backwards compatibility sake so the `string` type
+// also implements the `Label` Kind
+c = "a" + ""
+fill(column: c, value: 0)
+
+builtin add : (x: A, y: A) => A where A: Addable
+
+// This must still be allowed, however a naive implementation would first infer `A <=> Label("a")`
+// and then fail to unify `Label("a") <=> Label("b")`. If we keep treating `Label(..)` as a string except in cases where it unifies to a variable that actually has the `Label` type.
+add(x: "a", y: "b")
+
+builtin func : (opt: string) => int
+
+// This will work since `Label("a")` is a sub type of `string`
+func(opt: "a")
+
+// Possible extension where we only allow some specific labels to be passed in
+builtin func : (opt: "option1" | "option2") => int
```

Comment:
Is the idea here to make a sort of enum type? For things like the `method` parameter to the `quantile` function?

Why would we use labels for this other than its convenient? Meaning I don't see a use case yet of when a label used in the context of a column name would be constrained to a specific set.